flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableFactoryService.java

```diff
@@ -103,9 +106,52 @@
 	 * @param <T> factory class type
 	 * @return the matching factory
 	 */
-	public static <T> T find(Class<T> factoryClass, Map<String, String> propertyMap, ClassLoader classLoader) {
+	public static <T extends TableFactory> T find(
+			Class<T> factoryClass,
+			Map<String, String> propertyMap,
+			ClassLoader classLoader) {
 		Preconditions.checkNotNull(classLoader);
-		return findInternal(factoryClass, propertyMap, Optional.of(classLoader));
+		return findSingleInternal(factoryClass, propertyMap, Optional.of(classLoader));
+	}
+
+	/**
+	 * Finds all table factories of the given class and property map.
+	 *
+	 * @param factoryClass desired factory class
+	 * @param propertyMap properties that describe the factory configuration
+	 * @param <T> factory class type
+	 * @return all the matching factories
+	 */
+	public static <T extends TableFactory> List<T> findAll(Class<T> factoryClass, Map<String, String> propertyMap) {
+		return findAllInternal(factoryClass, propertyMap, Optional.empty());
```

Comment:
Hi, should we use Flink user classloader instead?  I package BlinkExecutorFactory into a fat  JAR, but it can't be found by Flink cluster.

Reply:
Hi @ssquan What is the exception you are getting? The `defaultLoader` actually already uses the user classloader. Could you make sure you have proper entries in the `META-INF/services/org.apache.flink.table.factories.TableFactory`. You might need a `https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ServicesResourceTransformer` if you have services coming from multiple dependencies.